File: 1335-maximum-candies-allocated-to-k-children/1335-maximum-candies-allocated-to-k-children.py
class Solution:
    def isPossible(self, candies, k, candy):
        child = sum(i // candy for i in candies)  # Count children served
        return child >= k
    def maximumCandies(self, candies: List[int], k: int) -> int:
        total_sum = sum(candies)
        max_candy = max(candies)

        if total_sum < k:
            return 0
        if total_sum == k:
            return 1

        l, r = 1, max_candy
        while l <= r:
            mid = (l + r) // 2
            if self.isPossible(candies, k, mid):
                l = mid + 1
            else:
                r = mid - 1

        return r

File: 1335-maximum-candies-allocated-to-k-children/1335-maximum-candies-allocated-to-k-children.java
class Solution {
    public boolean isPossible(int[] candies,long k,long candy){
        long child = 0;
        for(int i:candies){
            child += i / candy;
        }
        return (child >= k);
    }
    public int maximumCandies(int[] candies, long k) {
        long sum = 0;
        for(int i : candies){
            sum += i;
        }
        if(sum < k)
            return 0;
        if(sum == k)
            return 1;
        long l = 1,r = sum;
        while(l <= r){
            long mid = (l + r) / 2;
            if(isPossible(candies, k, mid)){
                l = mid + 1;
            }else{
                r = mid - 1;
            }
        }
        return (int)r;
    }
}